enterprise/trackedusers/users_reporter.go

```diff
@@ -0,0 +1,223 @@
+				logger.NewIntField("jobId", job.JobID))
+			continue
+		}
+
+		if sourceIDtoFilter != nil && sourceIDtoFilter[sourceID] {
+			u.log.Debug("source to filter", obskit.SourceID(sourceID))
+			continue
+		}
+		userID := gjson.GetBytes(job.EventPayload, "batch.0.userId").String()
+		anonymousID := gjson.GetBytes(job.EventPayload, "batch.0.anonymousId").String()
+
+		if userID == "" && anonymousID == "" {
+			u.log.Warn("both userID and anonymousID not found in job event payload", obskit.WorkspaceID(job.WorkspaceId),
+				logger.NewIntField("jobId", job.JobID))
+			continue
+		}
+
+		if workspaceSourceUserIdTypeMap[job.WorkspaceId] == nil {
+			workspaceSourceUserIdTypeMap[job.WorkspaceId] = make(map[string]map[string]*hll.Hll)
+		}
+
+		if userID != "" {
+			workspaceSourceUserIdTypeMap[job.WorkspaceId][sourceID] = u.recordIdentifier(workspaceSourceUserIdTypeMap[job.WorkspaceId][sourceID], userID, idTypeUserID)
+		}
+
+		if anonymousID != "" {
+			workspaceSourceUserIdTypeMap[job.WorkspaceId][sourceID] = u.recordIdentifier(workspaceSourceUserIdTypeMap[job.WorkspaceId][sourceID], anonymousID, idTypeAnonymousID)
+		}
+
+		if userID != "" && anonymousID != "" {
+			combinedUserIDAnonymousID := combineUserIDAnonymousID(userID, anonymousID)
+			workspaceSourceUserIdTypeMap[job.WorkspaceId][sourceID] = u.recordIdentifier(workspaceSourceUserIdTypeMap[job.WorkspaceId][sourceID], combinedUserIDAnonymousID, idTypeUserIDAnonymousIDCombination)
+		}
+	}
+
+	if len(workspaceSourceUserIdTypeMap) == 0 {
+		u.log.Warn("no data to collect", obskit.WorkspaceID(jobs[0].WorkspaceId))
+		return nil
+	}
+
+	reports := make([]*UsersReport, 0)
+	for workspaceID, sourceUserMp := range workspaceSourceUserIdTypeMap {
+		sourceIDReports := make([]*UsersReport, 0, len(sourceUserMp))
+		sourceIDReports = append(sourceIDReports, lo.MapToSlice(sourceUserMp, func(sourceID string, userIdTypeMap map[string]*hll.Hll) *UsersReport {
+			return &UsersReport{
+				WorkspaceID:              workspaceID,
+				SourceID:                 sourceID,
+				UserIDHll:                userIdTypeMap[idTypeUserID],
+				AnonymousIDHLL:           userIdTypeMap[idTypeAnonymousID],
+				IdentifiedAnonymousIDHLL: userIdTypeMap[idTypeUserIDAnonymousIDCombination],
+			}
+		})...)
+		reports = append(reports, sourceIDReports...)
+	}
+	return reports
+}
+
+func (u *UniqueUsersReporter) ReportUsers(ctx context.Context, reports []*UsersReport, tx *txn.Tx) error {
+	if len(reports) == 0 {
+		return nil
+	}
+	stmt, err := tx.PrepareContext(ctx, pq.CopyIn("tracked_users_reports",
+		"workspace_id",
+		"instance_id",
+		"source_id",
+		"reported_at",
+		"userid_hll",
+		"anonymousid_hll",
+		"identified_anonymousid_hll",
+	))
+	if err != nil {
+		return fmt.Errorf("preparing statement: %v", err)
+	}
+	defer func() { _ = stmt.Close() }()
+
+	for _, report := range reports {
+		_, err := stmt.Exec(report.WorkspaceID,
+			u.instanceID,
+			report.SourceID,
+			time.Now(),
```

Comment:
It is better to DI `time.Now` for testing:

```suggestion
			u.Now(),
```